File: C/99. Recover Binary Search Tree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
void recoverTree(struct TreeNode* root){
    struct TreeNode* f=NULL;
    struct TreeNode* s=NULL;
    struct TreeNode* pre=NULL;
    void helper(struct TreeNode* root){
        if(root==NULL){
            return;
        }
        helper(root->left);
        if(pre && pre->val>root->val){
            if(f==NULL){
                f=pre;
            }
            s=root;
        }
        pre=root;
        helper(root->right);
    }
    helper(root);
    int temp = f->val;
    f->val=s->val;
    s->val=temp;
}

File: Java/99. Recover Binary Search Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    TreeNode f=null;
    TreeNode s=null;
    TreeNode pre=null;
    public void recoverTree(TreeNode root) {
        helper(root);
        int temp = f.val;
        f.val=s.val;
        s.val=temp;
    }
    public void helper(TreeNode root){
        if(root==null){
            return;
        }
        helper(root.left);
        if(pre!=null && pre.val>root.val){
            if(f==null){
                f=pre;
            }
            s=root;
        }
        pre=root;
        helper(root.right);
    }
}
